crates/worker/src/partition/state_machine/lifecycle/event.rs

```diff
@@ -9,60 +9,127 @@
 use restate_types::identifiers::InvocationId;
-use restate_types::journal_v2::raw::RawEntry;
-use restate_types::journal_v2::{EntryMetadata, EntryType};
+use restate_types::journal_events::raw::RawEvent;
 use tracing::trace;
 
+pub struct OnInvokerEventCommand {
+    pub invocation_id: InvocationId,
+    pub invocation_status: InvocationStatus,
+    pub event: RawEvent,
+}
+
+impl<'ctx, 's: 'ctx, S: JournalEventsTable + InvocationStatusTable>
+    CommandHandler<&'ctx mut StateMachineApplyContext<'s, S>> for OnInvokerEventCommand
+{
+    async fn apply(self, ctx: &'ctx mut StateMachineApplyContext<'s, S>) -> Result<(), Error> {
+        let Self {
+            invocation_id,
+            mut invocation_status,
+            event,
+        } = self;
+        ApplyEventCommand {
+            invocation_id,
+            invocation_status: &mut invocation_status,
+            event,
+        }
+        .apply(ctx)
+        .await?;
+
+        // Store invocation status
+        ctx.storage
+            .put_invocation_status(&invocation_id, &invocation_status)
+            .await
+            .map_err(Error::Storage)?;
+
+        Ok(())
+    }
+}
+
 pub(super) struct ApplyEventCommand<'e> {
     pub(super) invocation_id: InvocationId,
     pub(super) invocation_status: &'e mut InvocationStatus,
-    pub(super) entry: &'e mut Option<RawEntry>,
+    pub(super) event: RawEvent,
 }
 
-impl<'e, 'ctx: 'e, 's: 'ctx, S: ReadOnlyJournalTable>
-    CommandHandler<&'ctx mut StateMachineApplyContext<'s, S>> for ApplyEventCommand<'e>
-{
-    async fn apply(self, ctx: &'ctx mut StateMachineApplyContext<'s, S>) -> Result<(), Error> {
+impl ApplyEventCommand<'_> {
+    async fn should_store_event<'e, 's: 'e, S: JournalEventsTable>(
+        journal_metadata: &JournalMetadata,
+        invocation_id: InvocationId,
+        new_event: &RawEvent,
+        storage: &'s mut S,
+    ) -> Result<bool, Error> {
         // --- Event deduplication logic
-        // Some Events are merged together when duplicated, to avoid storing too many duplicate events.
-        // This logic is completely optional and can fail silently.
-        let Some(journal_metadata) = self.invocation_status.get_journal_metadata() else {
-            return Ok(());
-        };
-        let last_entry_index = journal_metadata.length - 1;
-        let Some(last_entry) = ctx
-            .storage
-            .get_journal_entry(self.invocation_id, last_entry_index)
+        // Events get duplicated to avoid storing too many duplicate events.
+        if new_event.deduplication_hash().is_none() {
+            return Ok(true);
+        }
+        if journal_metadata.events == 0 {
+            return Ok(true);
+        }
+
+        let last_event_index = journal_metadata.events - 1;
+        let Some(last_event) = storage
+            .get_journal_event(invocation_id, last_event_index)
             .await?
         else {
-            return Ok(());
+            return Ok(true);
         };
```

Comment:
I'm assuming that this will be replaced by dudupe at proposal time in a later PR, no need for every node to keep deduping on replay. Is that correct?